lib/Twig/Environment.php

```diff
@@ -551,13 +551,20 @@ public function setCompiler(Twig_Compiler $compiler)
     /**
      * Compiles a node and returns the PHP code.
      *
-     * @param Twig_Node $node A Twig_Node instance
+     * @param Twig_Node   $node   A Twig_Node instance
+     * @param string|null $source The template source code
      *
      * @return string The compiled PHP source code
      */
-    public function compile(Twig_Node $node)
+    public function compile(Twig_Node $node, $source = null)
     {
-        return $this->getCompiler()->compile($node)->getSource();
+        $compiled = $this->getCompiler()->compile($node)->getSource();
+
+        if (isset($source[0])) {
+            $compiled .= '// '.str_replace(array("\r\n", "\r", "\n"), "\n// ", $source)."\n";
```

Comment:
Having a very long one liner is very problematic in some text editor that do not deal very well with wide texts.